### ISSUE TITLE
xRegistry: Unable to create binary registry value with data '0x00'

### ISSUE THREAD

**Issue by Amedama96** [PowerShell/xPSDesiredStateConfiguration#276]
## Description
I want to create binary registry value with data '0x00', but xRegistry resource set none value.

## Detail
I want to create binary reg value with data zero
![zero](https://cloud.githubusercontent.com/assets/18225518/20457978/7ce8e132-aedb-11e6-9db8-c02d2fb2b8ae.png)

For that, write and exec psdsc configuration like...
```PowerShell
xRegistry Zero_RegBinary
{
    Ensure = "Present"
    Key = "HKEY_LOCAL_MACHINE\SYSTEM\Tests"
    ValueName = "TestValue"
    ValueData = "0x00"    # same result if you set ValueData="0", 
    ValueType = "Binary"
}
```
Then, sadly, xRegistry resource create a reg value that has 'zero-length binary value'
![none](https://cloud.githubusercontent.com/assets/18225518/20458005/ff32c978-aedb-11e6-897a-d7529d558480.png)

 I confirmed the issue also have the In-box Regisrtry resource in Windows 10 1607.

## How to fix
I've discovered the issue caused by [this code](https://github.com/PowerShell/xPSDesiredStateConfiguration/blob/37043240c09220e134c76a427b2364b949c7e424/DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1#L1198).
```Powershell
$val = $Data[0].TrimStart('0x')
```
Apparently the code is intended to remove the first '0x' of $Data[0].
but if $Data[0] = '0' or '0x00', empty value is set in $val.
Perhaps this is undesired behavior.

I would like to propose to modify the code like...
```Powershell
if($Data[0].StartsWith('0x')) {
   $val = $Data[0].Substring(2)
}
else {
   $val = $Data[0]
}
```
It's a little redundant, but I think it is easy to understand and behave properly.

**Comment by LiamFisher92** [comment]
Hey @Amedama96, well done with finding this. I don't think the issue is as simple as taking a sub string of the string value. A value with '0x' as a prefix means that the value is a Hexadecimal value (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, A, B, C, D, E, F), so if the value contained any hexadecimal letters, [Int]::TryParse('F') would fail, and you would be in the same situation. The only way I know of getting round this is to use [System.Convert]::ToInt32() and specifying the base. 
Binary is Base 2, Integer (denary) is base 10, and Hexadecimal is base 16, so to convert a value of '0xF' to a base 10 integer you would use:
`[System.Convert]::ToInt32('0xF', 16)`
This will give you the true value of the integer! Hope this helps, let me know if you need any more help.

Liam

**Comment by LiamFisher92** [comment]
Hey @Amedama96, well done for finding this. After a little investigation, I found the real issue is that you can't assign a RegistryKey from a hexadecimal value. [System.Int32]::TryParse requires a string representation of an integer value. "0x00" is a string representation of a hexadecimal value. Fixing the parsing to allow for hexadecimal values would fix your problem. I've commented on the PR on how to fix this. Let me know if you need any help.

**Comment by Amedama96** [comment]
This issue fixed and merged #299 by @kwirkykat
